trellis/containers/transforms.hpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (C) 2022 Agtonomy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef TRELLIS_CONTAINERS_TRANSFORMS_HPP
+#define TRELLIS_CONTAINERS_TRANSFORMS_HPP
+
+#include <Eigen/Geometry>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "trellis/core/time.hpp"
+
+namespace trellis {
+namespace containers {
+
+/**
+ * Transforms a container to hold rigid coordinate frame transformations
+ *
+ * This class holds transforms with associated timestamps. It also maintains a validity window, which is a duration of
+ * time that the transform is valid for. Lookups are done based on a timestamp. The algorithm will first find the
+ * transform with the timestamp nearest the requested one. Then the validity window is used to determine if the
+ * timestamp is close enough in time. There is also a notion of forever in which a transform is always valid. This is
+ * useful for transforms that are known to be unchanging.
+ *
+ * Future improvements:
+ * - Support a graph structure to allow lookups that are transitive. Say we know A -> B -> C, we should support a lookup
+ * of A -> C
+ */
+class Transforms {
+ public:
+  /**
+   * Tranlsation represents a translation in 3D space
+   */
+  struct Translation {
+    double x;
+    double y;
+    double z;
+  };
+
+  /**
+   * Rotation represents a rotation in 3D space represented as a quaternion
+   */
+  struct Rotation {
```

Comment:
Similarly, you could replace Rotation with Eigen::Vector4d, which has accessor methods x(), y(), z(), and w(). 

Also, this struct should probably be called `Quaternion`.

I guess the core question is: why have these structs at all when the `Eigen` ones are already there and can be operated on directly?

Reply:
I went ahead and used the Eigen type here